### ISSUE TITLE
[FEATURE] Conan simplify build procedure

### ISSUE THREAD

**Issue by redradist** [conan-io/conan-center-index#6149]
**C and C++** known to be hard to work with, because of lots of reasons ...
One of the reason that user should know how to install particular toolchain and configure it

Such languages as **Rust** has **RustUp**, but we do not (:

What if Conan will add downloading, configuring and caching of toolchain ?!

Consider use-case:

```bash
conan install --toolchain clang-arm
conan build
```

It will simplify working with C and C++ ;)

Here is original topic:
https://github.com/conan-io/conan/issues/6835

**Comment by redradist** [comment]
There are already available some package for [llvm toolkit](https://github.com/conan-io/conan-center-index/tree/master/recipes/llvm-core/all), but it is not marked in special way and that it cannot be used in such workfolw:
```console
conan install --toolchain clang-arm
conan build
```
I think there should be special flag and official support from `conan` side to support such use-case ...

**Comment by jgsogo** [comment]
Hi, @redradist .

This is not something that belongs to this repository, here **we can try to provide packages with tools**, compilers and so on that will simplify our C++ developers' life, but we cannot add/remove arguments to Conan commands to add/remove functionality.

Given that some toolchain is available (required Conan packages) you can use them like it is recomended in  https://github.com/conan-io/conan/issues/6835 or you can use profiles for that. You could list your _toolchain_ like:

**profile.build**
```
[settings]
# Settings for your build machine

[build_requires] 
gcc/10.0.0
cmake/3.20
# more packages needed for your _toolchain_, _dev environment_

[env]
# You might need some environment variables: paths, custom CMake definitions,...
```

And then you can run:

```
conan install zlib/1.2.11@ --profile:host=<host-profile> --profile:build=profile.build --build
```

and it will use the _development environment_ you have declared in your build profile.

**Comment by redradist** [comment]
@jgsogo I am not taking about profiles, I am taking about:
1) Adding separate flag to package that indicates it is toolchain
2) Allow to manage installed toolchains (as conan package, but with special attributes/flags): add, update, remove
3) Allow for install toolchain generate profile, toolchain file for specific build system (toolchain.cmake, toolchain.make)

**Comment by ericriff** [comment]
I'm packaging my own toolchains for `arm` and `aarch64` and said packages expose the relevant env variables like CC, CXX, CPP, LD,CXX_FLAGS, SYSROOTS, etc.
Then I list one of those packages as `build_requirement` in a conan profile. 
When I build something using that profile, the build system sees the env variables and uses them to crosscompile.

**Comment by jgsogo** [comment]
@redradist , I think I understand what you mean, and with the current features it is something you can manage using profiles and build-requires. Maybe they are not 100% accurate to your needs, but probably they are close enough.

If you need new features in order to achieve what you need, then you really need to ask and get traction in the Conan repository, not here. **In `conan-center-index` we can only use what is already provided by the Conan client**, we can't add new things to the command line or modify the way Conan works.

I'm sorry, but this issue is beyond the scope of this repository. Here we can show you how we use existing packages with existing Conan features (thanks, Eric), or we can try to improve the packages to export other environment variables, compile more executables or add more configurations,...

**Comment by redradist** [comment]
@jgsogo 
> 
> 
> @redradist , I think I understand what you mean, and with the current features it is something you can manage using profiles and build-requires. Maybe they are not 100% accurate to your needs, but probably they are close enough.
> 
> If you need new features in order to achieve what you need, then you really need to ask and get traction in the Conan repository, not here. **In `conan-center-index` we can only use what is already provided by the Conan client**, we can't add new things to the command line or modify the way Conan works.
> 
> I'm sorry, but this issue is beyond the scope of this repository. Here we can show you how we use existing packages with existing Conan features (thanks, Eric), or we can try to improve the packages to export other environment variables, compile more executables or add more configurations,...

I agree that this feature should be discussed in `conan` repository, but in this issue I was sent here https://github.com/conan-io/conan/issues/6835

Also `build-requires` is close but not completly ...
To understand the difference you should look like it is done in `Rust` with `rustup` tool

**Comment by jgsogo** [comment]
I think you were forwarded here regarding the `gcc` and `clang` packages, or other packages that could be used in a toolchain... but this is not the repo to discuss new Conan features.

**Comment by redradist** [comment]
@jgsogo 
> 
> 
> I think you were forwarded here regarding the `gcc` and `clang` packages, or other packages that could be used in a toolchain... but this is not the repo to discuss new Conan features.

No only, also I was forwarded here to see if community is interested in such conan feature (see original issue)

**Comment by jgsogo** [comment]
If you have a look at the issues and interactions in this repository, you will find that almost all of them are related to recipes themselves, new package requests, new configurations to build,... and, whenever something is related to Conan client itself, we move the conversation there and link to it.

The community around `conan-center-index` is all about packages and recipes. For sure we welcome ideas that can improve Conan itself, and you can find in recipes here some evidence if those features are needed or not, and examples to use to prove your point... but the feature design, implementation details, user stories,... all of them should be exposed in  `conan-io/conan` repository. 

Here you will find that this issue will not get attention as many others arrive and it is not related to any concrete package.

**Comment by jgsogo** [comment]
The issue or feature request belongs to `conan-io/conan` repository. If it was declined there it makes no sense to talk about it here. If you want to suggest a dependency graph we should work on to get some packages into ConanCenter, that's ok, we can try to grab attention from this community and try to push all together into packaging the requirements.

**Comment by memsharded** [comment]
To clarify, and summarize my comment in https://github.com/conan-io/conan/issues/6835#issuecomment-898906483:

- The Conan feature is there, it is already possible to package compilers, toolchain in Conan packages and use them as ``build_requires``. 
- The effort is in creating those packages, often it is not easy to compile things like the compilers themselves.
- This is something community driven, these contents (compilers, toolchains) are third parties, not maintained, built, or released by Conan team.

**Comment by redradist** [comment]
@memsharded No really ...

This use-case is not covered by `conan`:
```console
conan install --toolchain clang-arm
```

It would require special flag `--toolchain` to be handled by `conan` and to add its support in `conan-center-index`

**Comment by redradist** [comment]
> 
> 
> To clarify, and summarize my comment in [conan-io/conan#6835 (comment)](https://github.com/conan-io/conan/issues/6835#issuecomment-898906483):
> 
>     * The Conan feature is there, it is already possible to package compilers, toolchain in Conan packages and use them as `build_requires`.
> 
>     * The effort is in creating those packages, often it is not easy to compile things like the compilers themselves.
> 
>     * This is something community driven, these contents (compilers, toolchains) are third parties, not maintained, built, or released by Conan team.

@memsharded As I understand `build_requires` will be installed only for build duration ?

What I suggest to have special flag in reciept that will say it is `toolchain` and to store it in special location like `rustup` does
